src/unix/lwt_domain.ml

```diff
@@ -0,0 +1,82 @@
+open Lwt.Infix
+
+module C = Domainslib.Chan
+module T = Domainslib.Task
+
+(* Maximum number of domains: *)
+let max_domains : int ref = ref 0
+
+let domains_count = ref 0
+
+let get_num_domains () = !max_domains
+
+(* Initial pool with only the parent domain *)
+let pool = ref (T.setup_pool ~num_additional_domains:0)
+
+let initialized = ref false
+
+(* Destroys old pool and creates a new pool with `num` domains *)
+let set_num_domains num =
+  initialized := true;
+  max_domains := num;
+  T.teardown_pool !pool;
+  pool := T.setup_pool ~num_additional_domains:(!max_domains - 1);
+  domains_count := !max_domains
```

Comment:
You set the current threashold at `<= 0`. What happens if I call this function with `1`?

Does it create zero other domains? In this case, what does `detach` do? Can it actually schedule the computation or does it create a promise that's never resolved? Does it schedule it on the same domain as the one running main.run?

Reply:
from a short experiment, setting max-domains to 1 seems to be blocking any call to `detach`

There are several options here:
- We reject `1` as well, to avoid blockage
- We special case `1` so that `detach` actually runs the computation on the same core
- other?

Reply:
You're right, setting `max_domains` to 1 blocks indefinitely since the bind waits for the detach computation to resolve and for the computation to run it needs the bind to succeed.

Rejecting `1` and making the minimum number of domains to be `2` seems the simpler way. We can still run this on one core, but the program would take longer to execute on one core. This means the program cannot be executed with one domain. If that's not an issue, we could go with the first option.

Reply:
On further consideration, I think we should give provision for running this with one domain. It can be useful for:
* Debugging - we've found debugging easier with one domain. It can also help classify concurrency bugs from other bugs.
* Benchmarking - running on one domain can act as the sequential baseline.

Reply:
I think `max_domains=1` should definitely be made to work in the Lwt_domains interface. A single core still counts as a multicore setup :) More generally, the domain running the Lwt engine should be able to run detached computations too, since that works already via Lwt_preemptive and systhreads.

Reply:
I don't think that we need to preemptive or systhread the detach computation if we have a single domain. I think that in this case we can just run the computation directly. If it is well documented I think the interface should be ok:

> If there is a single domain, then `detach f x` is just `return (f x)`. In other words, if there isn't a separate domain to execute the computation on, then the computation is run on the domain that also executes the main Lwt event loop.

I think this is better (than using systhreads) for the following reasons:
- When we eventually move away from Lwt_unix and into Lwt(core) it'll be easier
- The cost model is easier to understand
- The concurrency/parallelism model is easier to understand

Reply:
When we have `max_domains=1`, the computation runs directly. (see 951df84dd181185f3b24f0f4a099df3a441facf1)